chain-test/src/test/scala/org/bitcoins/chain/blockchain/ChainHandlerTest.scala

```diff
@@ -31,4 +41,96 @@ class ChainHandlerTest extends ChainUnitTest {
       foundHeaderF.map(found => assert(found.get == newValidHeader))
   }
 
+  it must "have an in-order seed" in { _ =>
+    val source =
+      scala.io.Source.fromURL(getClass.getResource("/block_headers.json"))
+    val arrStr = source.getLines.next
+    source.close()
+
+    import org.bitcoins.rpc.serializers.JsonReaders.BlockHeaderReads
+    val headersResult = Json.parse(arrStr).validate[Vector[BlockHeader]]
+    if (headersResult.isError) {
+      fail(headersResult.toString)
+    }
+
+    val blockHeaders = headersResult.get
+
+    blockHeaders.reduce[BlockHeader] {
+      case (prev, next) =>
+        assert(next.previousBlockHashBE == prev.hashBE)
+        next
+    }
+
+    succeed
+  }
+
+  it must "be able to process and fetch real headers from mainnet" in {
+    chainHandler: ChainHandler =>
+      val source =
+        scala.io.Source.fromURL(getClass.getResource("/block_headers.json"))
+      val arrStr = source.getLines.next
+      source.close()
+
+      import org.bitcoins.rpc.serializers.JsonReaders.BlockHeaderReads
+      val headersResult = Json.parse(arrStr).validate[Vector[BlockHeader]]
+      if (headersResult.isError) {
+        fail(headersResult.toString)
+      }
+
+      val firstPowChange = (FIRST_BLOCK_HEIGHT / chainHandler.chainParams.difficultyChangeInterval + 1) * chainHandler.chainParams.difficultyChangeInterval
+
+      val blockHeaders =
+        headersResult.get.drop((firstPowChange - FIRST_BLOCK_HEIGHT).toInt)
+
+      val firstBlockHeaderDb = BlockHeaderDbHelper.fromBlockHeader(firstPowChange - 2, ChainTestUtil.blockHeader562462)
+
+      val secondBlockHeaderDb =
+        BlockHeaderDbHelper.fromBlockHeader(firstPowChange - 1,
+                                            ChainTestUtil.blockHeader562463)
+
+      val thirdBlockHeaderDb =
+        BlockHeaderDbHelper.fromBlockHeader(firstPowChange,
+                                            ChainTestUtil.blockHeader562464)
+
+      chainHandler.blockchain.blockHeaderDAO
+        .createAll(Vector(firstBlockHeaderDb, secondBlockHeaderDb, thirdBlockHeaderDb))
+        .flatMap { _ =>
+          var processorF = Future.successful(
+            chainHandler.copy(blockchain = chainHandler.blockchain.copy(
+              headers = Vector(thirdBlockHeaderDb, secondBlockHeaderDb, firstBlockHeaderDb))))
+          var blockCount = firstPowChange
+
+          // Takes way too long to do all blocks
+          val blockHeadersToTest = blockHeaders.tail.take((2*chainHandler.chainParams.difficultyChangeInterval + 1).toInt)
```

Comment:
will this get better with the async version?

Reply:
What async version?